### PR TITLE
Feature/zenko 713 new management agent

### PR DESCRIPTION
**Pull request by jeremyds**
# Pull request template

### Motivation and context

Why is this change required? What problem does it solve?

### Related issues

Please use the following link syntaxes #600 to reference issues in the
current repository

### Add tests to cover the changes

New tests added or existing tests modified to cover all changes

### Sign your work

In order to contribute to the project, you must sign your work
https://github.com/scality/Guidelines/blob/master/CONTRIBUTING.md#sign-your-work

Thank you again for contributing! We will try to test and integrate the change
as soon as we can.

### CODE REVIEW

**Comment by jeremyds** [comment]
for reviewers: only the last commit (ZENKO 713) is relevant, others are in [this PR.](https://github.com/scality/S3/pull/1378)

File: managementAgent.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    });
+
+    wss.on('error', error => {
+        logger.error('management agent websocket server error', { error });
+    });
+
+    addOverlayMessageListener((remoteOverlay) => {
+        wss.clients.forEach((client) => {
+            if (client.readyState !== client.OPEN) {
+                logger.warning('client socket not in ready state', { client });
```
Comment (jeremyds) [inline managementAgent.js]: logger.error

File: lib/management/configuration.js [hunk 2]
```diff
@@ -13,6 +13,9 @@ const managementDatabaseName = 'PENSIEVE';
 const replicatorEndpoint = 'zenko-cloudserver-replicator';
 const { decryptSecret } = arsenal.pensieve.credentialUtils;
 
+
+let configurationListeners = [];
```
Comment (rachedbenmustapha) [inline lib/management/configuration.js]: This can be const if it's pushed into

File: managementAgent.js [hunk 3]
```diff
@@ -3,8 +3,16 @@ const Uuid = require('uuid');
 
 const logger = require('./lib/utilities/logger');
 const { initManagement } = require('./lib/management');
+const { addConfigurationListener } = require('./lib/management/configuration');
 const _config = require('./lib/Config').config;
+const WebSocket = require('ws');
+
+
+//TODO: signal handler to cleanly shutdown server
+//TODO: ping messages between server and client
```
Comment (rachedbenmustapha) [inline managementAgent.js]: I think it would be good to implement the ping part in this PR

**Review by rachedbenmustapha — CHANGES_REQUESTED** [review]
In order to be more resilient to disconnections and timing issues, we should also send the latest overlay on the websocket as soon as a client is connected.

File: lib/management/configuration.js [hunk 4]
```diff
@@ -300,10 +310,22 @@ function loadCachedOverlay(log, callback) {
         });
 }
 
+/**
```
Comment (jeremyds) [inline lib/management/configuration.js]: should be removed, interactive rebase error I guess, now in push.js.

File: managementAgent.js [hunk 5]
```diff
@@ -55,5 +134,19 @@ function managementAgentWS() {
                 }
             });
         });
-    });
+    }
+
+    checkBrokenConnections() {
+        this.wss.clients.forEach((client) => {
+            if (!client.isAlive) {
+                logger.error('close broken connection');
```
Comment (rachedbenmustapha) [inline managementAgent.js]: This should be info instead of error - errors are for abnormal conditions and a client restarting is normal.

File: lib/management/configuration.js [hunk 6]
```diff
@@ -256,12 +259,19 @@ function saveConfigurationVersion(cachedOverlay, remoteOverlay, log, cb) {
         metadata.putObjectMD(managementDatabaseName, objName, remoteOverlay,
             {}, log, error => {
                 if (error) {
-                    log.error('could not save configuration version',
+                    log.error('could not save configuration',
```
Comment (rachedbenmustapha) [inline lib/management/configuration.js:259]: It's not new with this change but since there's already a diff here: we should not continue to update the version pointer if there was a failure saving the overlay in the first place.

File: managementAgent.js [hunk 7]
```diff
@@ -1,51 +1,130 @@
...
-    wss.on('listening', ws => {
-        logger.info('management agent websocket server listening', { port });
-    });
+        socket.isAlive = true;
+        socket.on('pong', hearthbeat.bind(socket));
 
-    wss.on('error', error => {
-        logger.error('management agent websocket server error', { error });
-    });
+        if (!socket.readyState === socket.OPEN) {
```
Comment (rachedbenmustapha) [inline managementAgent.js]: Cosmetic but it took me a few seconds to see that the condition is reversed, I suggest using `!==` instead

File: managementAgent.js [hunk 8]
```diff
@@ -1,51 +1,130 @@
...
+        /* Define REPORT_TOKEN env variable needed by the management
+         * module. */
+        process.env.REPORT_TOKEN = process.env.REPORT_TOKEN
+          || _config.reportToken
+          || Uuid.v4();
+
+        metadata.setup(error => {
+            if (error) {
+                const msg = `failed to setup metadata: ${error}`;
+                throw new Error(msg);
```
Comment (rachedbenmustapha) [inline managementAgent.js]: We should not throw in async callbacks, instead call a callback with the failure so that the toplevel code can schedule a retry

File: managementAgent.js [hunk 9]
```diff
@@ -1,51 +1,130 @@
...
+            }
+            initManagement(logger.newRequestLogger(), (overlay) => {
+                if (overlay) {
+                    this.loadedOverlay = overlay;
+                    this.startServer();
+                } else {
+                    const msg = `failed to init management: ${error}`;
+                    throw new Error('failed to init management');
```
Comment (rachedbenmustapha) [inline managementAgent.js]: Same here, we should not throw

File: managementAgent.js [hunk 10]
```diff
@@ -1,51 +1,130 @@
...
+        logger.info('websocket server listening',
+                    { port: this.port });
+    }
+
+    onError(error) {
+        logger.error('websocket server error', { error });
+    }
+
+    onNewOverlay(remoteOverlay) {
+        //XXX patch configuration?
```
Comment (rachedbenmustapha) [inline managementAgent.js]: I don't think the server has anything to patch as the overlay only contains s3-protocol specific items. On thing that will come later is the `browserAccess` flag

File: managementAgent.js [hunk 11]
```diff
@@ -1,51 +1,130 @@
...
-        logger.info('management agent websocket server listening', { port });
-    });
+        socket.isAlive = true;
+        socket.on('pong', hearthbeat.bind(socket));
 
-    wss.on('error', error => {
-        logger.error('management agent websocket server error', { error });
-    });
+        if (!socket.readyState === socket.OPEN) {
+            logger.error('client socket not in ready state', { socket });
```
Comment (rachedbenmustapha) [inline managementAgent.js]: This could include a lot of unpleasant data to look at that can be big or circular, so for clarity and for preserving performance we should select a few meaningful attributes (if available) such as `host`, `port`, `readyState` and so on

File: lib/management/agentClient.js [hunk 12]
```diff
@@ -0,0 +1,16 @@
+const managementAgentMessageType = {
+    /** Message that contains a configuration overlay */
+    NEW_OVERLAY_VERSION: 1,
+    /** Message that contains the loaded overlay */
+    LOADED_OVERLAY: 2,
```
Comment (rachedbenmustapha) [inline lib/management/agentClient.js]: Why is there a different message type `LOADED_OVERLAY`? I think only one message type to push overlays would be clearer.

**Review by rachedbenmustapha — CHANGES_REQUESTED** [review]

File: lib/management/agentClient.js (hunk 12 above)
Comment (jeremyds) [reply lib/management/agentClient.js]: I thought it maybe useful to distinguish the loaded overlay from a new one. But as long as I don't have a precise use case in mind, let's have only one message type for the moment.

File: lib/management/configuration.js (hunk 2 above)
Comment (jeremyds) [reply lib/management/configuration.js]: not relevant anymore, code has been moved to push.js

File: lib/management/configuration.js (hunk 6 above)
Comment (jeremyds) [reply lib/management/configuration.js:259]: right, I am going to fix it in a separate commit.

File: managementAgent.js (hunk 8 above)
Comment (jeremyds) [reply managementAgent.js]: done

File: managementAgent.js (hunk 9 above)
Comment (jeremyds) [reply managementAgent.js]: done too

File: managementAgent.js (hunk 7 above)
Comment (jeremyds) [reply managementAgent.js]: right, not easy to read

File: managementAgent.js (hunk 11 above)
Comment (jeremyds) [reply managementAgent.js]: done

File: managementAgent.js (hunk 10 above)
Comment (jeremyds) [reply managementAgent.js]: thanks for the answer, XXX to remove

File: managementAgent.js (hunk 5 above)
Comment (jeremyds) [reply managementAgent.js]: done

File: managementAgent.js [hunk 13]
```diff
@@ -1,59 +1,159 @@
...
+          || _config.reportToken
+          || Uuid.v4();
 
-setTimeout(() => {
-    managementAgentWS();
-}, 6);
+        metadata.setup(error => {
+            if (error) {
+                const msg = `failed to setup metadata: ${error}`;
+                return cb(new Error(msg));
```
Comment (rachedbenmustapha) [inline managementAgent.js]: Real `Error` objects usually don't print well in our json logs so just forwarding the error we get to the callback is safer. A log before that saying `log.error('failed to setup metadata', { error })` would be enough to track

File: lib/management/push.js [hunk 14]
```diff
@@ -141,10 +146,7 @@ function startWSManagementClient(pushEndpoint, instanceId, token) {
               || 'localhost';
             const port = process.env.SECURE_CHANNEL_DEFAULT_FORWARD_TO_PORT
               || _config.port;
-            socket = net.createConnection({
-                host: host,
-                port: port,
-            });
+            socket = net.createConnection(host, port);
```
Comment (rachedbenmustapha) [inline lib/management/push.js]: The order is wrong here, its port then host (nodejs nicety). Why the change?

**Review by rachedbenmustapha — CHANGES_REQUESTED** [review]

File: lib/management/index.js [hunk 15]
```diff
@@ -80,14 +82,14 @@ function initManagement(log) {
...
         (instanceId, cb) => initManagementCredentials(
             managementEndpoint, instanceId, log, cb),
-        (instanceId, token, cb) => loadCachedOverlay(log, (err, conf) => {
+        (instanceId, token, cb) => loadCachedOverlay(log, (err, overlay) => {
             if (err) {
                 return cb(err);
             }
-            return patchConfiguration(conf, log,
-                err => cb(err, instanceId, token));
+            return patchConfiguration(overlay, log,
```
Comment (nicolas2bert) [inline lib/management/index.js]: It might be clearer to add another waterfall callback here for `patchConfiguration()`. 
`(overlay, instanceId, ... cb) => patchConfiguration(overlay...)`

File: managementAgent.js [hunk 16]
```diff
@@ -1,59 +1,159 @@
...
+
+
+// TODO: auth?
+
+const CHECK_BROKEN_CONNECTIONS_FREQUENCY_MS = 15000;
+
+
+class ManagementAgentServer {
+    constructor() {
+        this.port = _config.managementAgent.port || 8010;
```
Comment (nicolas2bert) [inline managementAgent.js]: Not sure how `_config.managementAgent.port` can be `undefined` or `0` since you already asserted it in the `Config.js`. _Better safe than sorry?_

File: managementAgent.js [hunk 17]
```diff
@@ -1,59 +1,159 @@
...
+        this.wss = null;
+        this.loadedOverlay = null;
+
```
Comment (nicolas2bert) [inline managementAgent.js]: You could also have done it this way:
```
this.stop = this.stop.bind(this);

process.on('SIGINT', this.stop);

process.on('SIGHUP', this.stop);

process.on('SIGQUIT', this.stop);

process.on('SIGTERM', this.stop);
```

**Review by nicolas2bert — APPROVED** [review]

File: managementAgent.js (hunk 16 above)
Comment (jeremyds) [reply managementAgent.js]: yes, exactly

File: lib/management/index.js (hunk 15 above)
Comment (jeremyds) [reply lib/management/index.js]: right, done

File: lib/management/push.js (hunk 14 above)
Comment (jeremyds) [reply lib/management/push.js]: wow... linter complains about the object form {port: value, host: value}

File: managementAgent.js (hunk 13 above)
Comment (jeremyds) [reply managementAgent.js]: done

File: managementAgent.js (hunk 3 above)
Comment (jeremyds) [reply managementAgent.js]: done (I missed this comment)

File: lib/management/push.js (hunk 14 above)
Comment (rachedbenmustapha) [reply lib/management/push.js]: Maybe it wants `{ host, port }` since the vars have the same name as the keys

File: managementAgent.js [hunk 18]
```diff
@@ -1,59 +1,169 @@
...
+                return;
+            }
+            client.isAlive = false;
+            client.ping();
+        });
+    }
 }
+
+const server = new ManagementAgentServer();
+server.start();
```
Comment (rachedbenmustapha) [inline managementAgent.js:162]: If `ManagementAgentServer.start()` fails and we don't provide a callback that exits, are we sure that the process will exit with an error status? When deployed in k8s for example if mongodb is not yet available we want to be sure that we retry initializing later. Arguably we could crash as well but it only works well when deployed inside an orchestrator.

**Review by rachedbenmustapha — CHANGES_REQUESTED** [review]

**Review by rachedbenmustapha — APPROVED** [review]
